Big endian fixes for dotnet runtime

nealef opened the pull request:
A series of fixes that enable big endian running of the dotnet command. With these fixes and fixes in https://github.com/dotnet/runtimelab/pull/650 the s390x run of the test suite results in 240 of 240 tests passing. 

1. System.Private.CoreLib/src/System/Text/ASCIIUtility.cs 
-- Big endian handling of widening to UTF16
2. System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Helpers.cs
-- Correct big endian extraction of UTF8 from two/four byte sequences
-- Correct detection of overlong rep for BE
3. System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Transcoding.cs
-- Correct transcoding to UTF8 for BE
4. System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatterWriter.cs
-- Remove debug message as we have run and verified the code
5. System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryParser.cs
-- Remove debug message as we have run and verified the code
6. System.Text.Encoding.CodePages/src/System.Text.Encoding.CodePages.csproj
-- Include reference for use in BE code
7. System.Text.Encoding.CodePages/src/System/Text/BaseCodePageEncoding.cs
-- Add code for BE reading of file header, code page header, and code page index
-- Remove debug message as we have run and verified the code
8. System.Text.Encoding.CodePages/src/System/Text/BaseCodePageEncoding.netcoreapp.cs
-- Add code for BE reading of code page index
9. System.Text.Encoding.CodePages/src/System/Text/DBCSCodePageEncoding.cs
-- Provide endian aware character reading
10. System.Text.Encoding.CodePages/src/System/Text/SBCSCodePageEncoding.cs
-- Endian aware code to map unicode to bytes

These fixes also address the problem described in issue https://github.com/dotnet/runtime/issues/44805.

@eiriktsarpalis @layomia 

Suggestion: area-System.Text.Encodings

Commit:
Big endian fixes for dotnet runtime

nealef commented:
I note failures such as:
```
Check failure on line 103 in src/libraries/System.Text.Encoding.CodePages/src/System/Text/BaseCodePageEncoding.cs

Azure Pipelines
/ runtime-dev-innerloop (Build Linux x64 release SourceBuild)
src/libraries/System.Text.Encoding.CodePages/src/System/Text/BaseCodePageEncoding.cs#L103
src/libraries/System.Text.Encoding.CodePages/src/System/Text/BaseCodePageEncoding.cs(103,79): error CS0246: (NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'Span<>' could not be found (are you missing a using directive or an assembly reference?)
```
However, I had built runtime on my own 64-bit system with this patches and got a clean bill:
```
Build succeeded.
    0 Warning(s)
    0 Error(s)
```
I am not sure why this is happening.

stephentoub commented:
> I am not sure why this is happening.

That project builds for various configurations that don't have a built-in Span:
https://github.com/dotnet/runtime/blob/45287ed696a4bea906cc15a4bc870960a4b0e623/src/libraries/System.Text.Encoding.CodePages/src/System.Text.Encoding.CodePages.csproj#L5

You could try adding a reference to System.Memory.

nealef commented:
> > I am not sure why this is happening.
> 
> That project builds for various configurations that don't have a built-in Span:
> https://github.com/dotnet/runtime/blob/45287ed696a4bea906cc15a4bc870960a4b0e623/src/libraries/System.Text.Encoding.CodePages/src/System.Text.Encoding.CodePages.csproj#L5
> 
> You could try adding a reference to System.Memory.

Thanks, what about things like: 
```
src\libraries\System.Text.Encoding.CodePages\src\System\Text\BaseCodePageEncoding.cs(4,14): error CS0234: (NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'Buffers' does not exist in the namespace 'System' (are you missing an assembly reference?)
```
That seems a bit harder to fix if that reference isn't present.

stephentoub commented:
> That seems a bit harder to fix if that reference isn't present.

That's why you'd add the reference :)  System.Memory provides the BinaryPrimitives type.

nealef commented:
> > That seems a bit harder to fix if that reference isn't present.
> 
> That's why you'd add the reference :) System.Memory provides the BinaryPrimitives type.

And dispense with that using `System.Buffers.Binary` altogether? [Excuse my naivety]

stephentoub commented:
> And dispense with that using System.Buffers.Binary altogether? [Excuse my naivety]

No.  System.Memory.dll includes the System.Buffers.Binary.BinaryPrimitives type.  You would need to add something like:
```
  <ItemGroup Condition="!$(TargetFramework.StartsWith('$(NetCoreAppCurrent)'))">
    <PackageReference Include="System.Memory" Version="$(SystemMemoryVersion)" />
  </ItemGroup>
```
to the project file, to include a package reference to the System.Memory package for the targets that don't have System.Memory in-box.

That's not going to solve all of the issues you're hitting, though.  For example, you've written code to use an overload of Stream.Read that simply doesn't exist in those previous versions.

nealef commented:
> That's not going to solve all of the issues you're hitting, though. For example, you've written code to use an overload of Stream.Read that simply doesn't exist in those previous versions.

Thanks again. I'll add those lines. As I, presumably, can't emulate that previous build environment is it acceptable to amend this PR to generate another around of building and debug from that or is that a gratuitous use of the CI environment?

stephentoub commented:
> As I, presumably, can't emulate that previous build environment

You can.  When you build from root, you're just building one specific target.  You can either build the whole repo for multiple targets (e.g. `build.cmd -allconfigurations`), or you can cd into the relevant directory here (`cd src\libraries\System.Text.Encoding.CodePages\src`) and do a build just of that project (`dotnet build`) which will build all of the configurations.

nealef commented:
Perfect! Thanks yet again.

Commit:
Enable endian fixes to build on older levels
- Use Stream.Read(byte[] target, int offset, int length) instead of Stream.Read(Span)
- Include System.Memory for older build levels
- Enable BaseCodePageEncoding.netcoreapp.cs to use the Stream.Read(Span) method by giving it it's own ReadCodePageIndex() method

nealef commented:
> That's not going to solve all of the issues you're hitting, though. For example, you've written code to use an overload of Stream.Read that simply doesn't exist in those previous versions.

I found those errors to which you referred and have pushed some changes. I got a clean build using the `dotnet build` in that directory. I then tried a build using the `-allconfigurations` option and it built for sometime before failing here:
```
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/Synthesis/ISsmlParser.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/Synthesis/SsmlParser.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/Synthesis/TtsEvent.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/Synthesis/TtsVoice.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/SrgsCompiler/SrgsCompiler.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/Synthesis/TtsEngineProxy.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/Synthesis/ISsmlParser.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/Synthesis/SsmlParser.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/Synthesis/TtsEngineProxy.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/Synthesis/TtsEvent.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/Synthesis/TtsVoice.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
CSC : error CS2001: Source file '/home/neale/runtime/src/libraries/System.Speech/src/Internal/SrgsCompiler/SrgsCompiler.cs' could not be found. [/home/neale/runtime/src/libraries/System.Speech/src/System.Speech.csproj]
```
I am making the probably rash assumption that this is due to something missing on my build machine. I'll await the latest checks to see how it goes there.

jkotas commented on src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatterWriter.cs:
```diff
@@ -268,7 +268,7 @@ private void WriteArrayAsBytes(Array array, int typeLength)
                 {
                     // we know that we are writing a primitive type, so just do a simple swap
-                    Debug.Fail("Re-review this code if/when we start running on big endian systems");
+                    // Debug.Fail("Re-review this code if/when we start running on big endian systems");
```
Just delete it instead of commenting it out

nealef replied:
Pushed amendment

nealef committed:
Delete Debug review message

GrabYourPitchforks commented on src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Transcoding.cs:
```diff
@@ -1134,7 +1134,7 @@ public static OperationStatus TranscodeToUtf8(char* pInputBuffer, int inputLengt
                     else
                     {
-                        pOutputBuffer[0] = (byte)(thisDWord >> 24); // extract [ AA 00 ## ## ]
+                        pOutputBuffer[0] = (byte)(thisDWord >> 16); // extract [ 00 AA ## ## ]
```
Thanks for correcting the comment as well! :)

nealef replied:
:-)

GrabYourPitchforks commented on src/libraries/System.Text.Encoding.CodePages/src/System/Text/SBCSCodePageEncoding.cs:
```diff
@@ -32,6 +33,18 @@ public SBCSCodePageEncoding(int codePage, int dataCodePage) : base(codePage, dat
         {
         }
 
+        internal static unsafe ushort ReadUInt16(byte *pByte)
```
Nit: standard library spacing is to use `T* param` instead of `T *param`. (Same comment on lines 40 and 44: remove the whitespace.)

nealef replied:
Updated and pushed

GrabYourPitchforks commented on src/libraries/System.Text.Encoding.CodePages/src/System/Text/SBCSCodePageEncoding.cs:
```diff
@@ -91,16 +104,16 @@ protected override unsafe void LoadManagedCodePage()
-                    char* pTemp = (char*)pBuffer;
                     for (int b = 0; b < 256; b++)
                     {
+                        char c = (char)ReadUInt16(pBuffer + 2*b);
```
Nit: `pBuffer + 2 * b`, or `&pBuffer[2 * b]`. (Insert whitespace around arithmetic operators.)

nealef committed:
Conform to style conventions

GrabYourPitchforks commented:
/azp run runtime

GrabYourPitchforks commented:
Previous CI run lost the artifacts somehow, so kicked off a new run. Hoping things will pass so that we can merge this in today.

GrabYourPitchforks commented:
The _Libraries Build windows allConfigurations x64 Debug_ failure is a known issue and the build engineering team is on it. I'll ignore it for now while waiting for the rest to finish.

GrabYourPitchforks commented:
Got it in. Thanks @nealef for your contribution! 🚀

nealef commented:
Thanks! I think there'll be another installment after some more testing.
